### PR TITLE
Custom protocol

### PR DESCRIPTION
**Pull request by milanzlicic**

### CODE REVIEW

File: src/service/PublishingService.cpp [hunk 1]
```diff
@@ -35,44 +38,90 @@ PublishingService::PublishingService(ConnectivityService& connectivityService,
...
+    m_buffer.notify();
+}
+
+PublishingService::State::State(PublishingService& service) : m_service{service} {}
+
+void PublishingService::DisconnectedState::run()
+{
+    while (m_service.m_run && !m_service.m_connected && !m_service.m_buffer.isEmpty())
+    {
+        const auto message = m_service.m_buffer.pop();
```
Comment (nanavuletic) [inline src/service/PublishingService.cpp:83]: Null check this message.

File: src/service/PublishingService.cpp [hunk 2]
```diff
@@ -35,44 +38,90 @@ PublishingService::PublishingService(ConnectivityService& connectivityService,
...
+    m_service.m_buffer.swapBuffers();
+}
 
-        std::unique_lock<std::mutex> locker{m_lock};
-        m_condition.wait(locker);
+void PublishingService::ConnectedState::run()
+{
+    while (m_service.m_run && m_service.m_connected && !m_service.m_buffer.isEmpty())
+    {
+        const auto message = m_service.m_buffer.pop();
```
Comment (nanavuletic) [inline src/service/PublishingService.cpp:97]: Null check this message.

File: src/service/PublishingService.cpp [hunk 3]
```diff
@@ -35,44 +38,90 @@ PublishingService::PublishingService(ConnectivityService& connectivityService,
...
+            LOG(ERROR) << "Failed to publish message";
+            break;
+        }
+    }
+
+    // publish persisted unitl new message arrives
+    while (m_service.m_run && m_service.m_connected && !m_service.m_persistence->empty() &&
+           m_service.m_buffer.isEmpty())
+    {
+        const auto message = m_service.m_persistence->front();
```
Comment (nanavuletic) [inline src/service/PublishingService.cpp:113]: Null check this message.

**Review by nanavuletic — CHANGES_REQUESTED** [review]

File: src/persistence/filesystem/GatewayFilesystemPersistence.cpp [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+            }
+
+            std::copy(files.begin(), files.end(), std::back_inserter(m_readingFiles));
+        }
+    }
+    else
+    {
+        if (!FileSystemUtils::createDirectory(m_persistPath))
+        {
+            LOG(ERROR) << "Could not create presist directory: " << m_persistPath;
```
Comment (srdjastankovic) [inline src/persistence/filesystem/GatewayFilesystemPersistence.cpp:140]: typo `presist` rename into `persist`

File: src/persistence/filesystem/MessagePersister.cpp [hunk 5]
```diff
@@ -0,0 +1,35 @@
+#include "MessagePersister.h"
```
Comment (srdjastankovic) [inline src/persistence/filesystem/MessagePersister.cpp:1]: Copyright

File: src/service/PublishingService.cpp [hunk 6]
```diff
@@ -35,44 +38,90 @@ PublishingService::PublishingService(ConnectivityService& connectivityService,
...
 
-        std::unique_lock<std::mutex> locker{m_lock};
-        m_condition.wait(locker);
+void PublishingService::ConnectedState::run()
+{
+    while (m_service.m_run && m_service.m_connected && !m_service.m_buffer.isEmpty())
+    {
+        const auto message = m_service.m_buffer.pop();
+        if (m_service.m_connectivityService.publish(message))
+        {
```
Comment (srdjastankovic) [inline src/service/PublishingService.cpp:99]: is this legal expression? do you have to insert continue?

File: src/service/PublishingService.cpp [hunk 7]
```diff
@@ -35,44 +38,90 @@ PublishingService::PublishingService(ConnectivityService& connectivityService,
...
+        }
+        else
+        {
+            m_service.m_persistence->push(message);
+            LOG(ERROR) << "Failed to publish message";
+            break;
+        }
+    }
+
+    // publish persisted unitl new message arrives
```
Comment (srdjastankovic) [inline src/service/PublishingService.cpp:109]: typo `unitl`

**Review by srdjastankovic — CHANGES_REQUESTED** [review]

File: src/WolkBuilder.cpp [hunk 8]
```diff
@@ -209,6 +223,12 @@ std::unique_ptr<Wolk> WolkBuilder::build()
         throw std::logic_error("Url downloader must be provided when url download is enabled");
     }
 
+    if ((m_dataProtocol == nullptr && m_statusProtocol != nullptr) ||
```
[the remark this answers is not in the document]
Comment (nanavuletic) [reply src/WolkBuilder.cpp:226]: Read the code below, the case where they are null is handled, this is just a prevention of one being null, and the other one not. They can be null, they just both must be null.

[the remark this answers is not in the document]
Comment (srdjastankovic) [reply src/WolkBuilder.cpp:226]: @nanavuletic than the log msg is wrong!

[the remark this answers is not in the document]
Comment (milanzlicic) [reply src/WolkBuilder.cpp:226]: If they are both null the defaults are used

**Review by nanavuletic — APPROVED** [review]

**Pull request merged**